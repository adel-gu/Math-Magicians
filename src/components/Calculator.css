button {
  border: none;
  padding: 15px 10px;
  background-color: aliceblue;
}

.calculator {
  width: 450px;
  max-width: 90%;
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  gap: 1px;
  box-shadow: 10px 8px 28px -7px rgba(0, 0, 0, 0.15);
}

.screen {
  background: var(--clr-screen);

  grid-column: span 4;
  padding: 5px;
  text-align: right;
}

.numberZero {
  grid-column: span 2;
}

.orang {
  background-color: var(--clr-orange);
  color: white;
}

button:hover {
  cursor: pointer;
  background: rgba(0, 0, 0, 0.1);
}

button.orang:hover {
  background-color: rgba(245, 145, 62, 0.8);
}
