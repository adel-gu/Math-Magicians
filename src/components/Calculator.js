import '../App.css';

const Screen = (props) => {
  return <span className='screen'>{props.result}</span>;
};

const Oprand = ({ data, className }) => {
  return <button className={className}>{data}</button>;
};

const Number = ({ value, className }) => {
  return <button className={className}>{value}</button>;
};

// Calculator

const Calculator = () => {
  return (
    <div className='calculator'>
      <Screen result='0' />
      {/* First row */}
      <Oprand data='AC' />
      <Oprand data='+/-' />
      <Oprand data='%' />
      <Oprand data='+' className='orang' />
      {/* Seccond row */}
      <Number value='7' />
      <Number value='8' />
      <Number value='9' />
      <Oprand data='*' className='orang' />
      {/* 3rd row */}
      <Number value='4' />
      <Number value='5' />
      <Number value='6' />
      <Oprand data='-' className='orang' />
      {/* 4th row */}
      <Number value='1' />
      <Number value='2' />
      <Number value='3' />
      <Oprand data='+' className='orang' />
      {/* 5th row */}
      <Number value='0' className='numberZero' />
      <Number value='.' />
      <Oprand data='=' className='orang' />
    </div>
  );
};

export default Calculator;
