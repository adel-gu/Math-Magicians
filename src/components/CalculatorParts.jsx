import PropTypes from 'prop-types';
// Sub Components => {Screen, oprands, Numbers}
const Screen = ({ result }) => <span className="screen">{result}</span>;
Screen.propTypes = { result: PropTypes.string.isRequired };

// Oprands buttons components
const Oprand = ({ data, className, onclick}) => (
  <button type="button" className={className} onClick={onclick}>
    {data}
  </button>
);
Oprand.propTypes = { data: PropTypes.string.isRequired };
Oprand.propTypes = { className: PropTypes.string };
Oprand.propTypes = { onclick: PropTypes.func.isRequired };

// Numbers buttons components
const Number = ({ value, className , onclick}) => (
  <button type="button" className={className} onClick={onclick}>
    {value}
  </button>
);
Number.propTypes = { value: PropTypes.string.isRequired };
Number.propTypes = { className: PropTypes.string};
Number.propTypes = { onclick: PropTypes.func.isRequired};

export {Screen, Oprand, Number};