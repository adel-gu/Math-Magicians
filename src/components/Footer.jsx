const Footer = () => {
  const mb = {
    marginBottom: '10px',
  };
  return (
    <footer className="footer">
      <div className="container header-footer py-2" style={mb}>
        <span>
          Build by
          <a href="https://github.com/adel-gu">Adel Guitoun ❤️</a>
        </span>
      </div>
    </footer>
  );
};

export default Footer;
