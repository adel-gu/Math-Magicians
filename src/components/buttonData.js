const buttonsData = [
  { data: 'AC', columnPos: 1, rowPos: 2 },
  { data: '+/-', columnPos: 2, rowPos: 2 },
  { data: '%', columnPos: 3, rowPos: 2 },
  {
    data: '+',
    columnPos: 4,
    rowPos: 2,
    color: 'orang',
  },
  { data: '7', columnPos: 1, rowPos: 3 },
  { data: '8', columnPos: 2, rowPos: 3 },
  { data: '9', columnPos: 3, rowPos: 3 },
  {
    data: 'x',
    columnPos: 4,
    rowPos: 3,
    color: 'orang',
  },
  { data: '4', columnPos: 1, rowPos: 4 },
  { data: '5', columnPos: 2, rowPos: 4 },
  { data: '6', columnPos: 3, rowPos: 4 },
  {
    data: '-',
    columnPos: 4,
    rowPos: 4,
    color: 'orang',
  },
  { data: '1', columnPos: 1, rowPos: 5 },
  { data: '2', columnPos: 2, rowPos: 5 },
  { data: '3', columnPos: 3, rowPos: 5 },
  {
    data: '/',
    columnPos: 4,
    rowPos: 5,
    color: 'orang',
  },
  {
    data: '0',
    columnPos: 1,
    rowPos: 6,
    width: 2,
  },
  { data: '.', columnPos: 3, rowPos: 6 },
  {
    data: '=',
    columnPos: 4,
    rowPos: 6,
    color: 'orang',
  },
];

export default buttonsData;
