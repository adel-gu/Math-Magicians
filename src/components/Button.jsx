import React from 'react';
import PropTypes from 'prop-types';

// Stylesheets
import '../styles/Calculator.css';

export const Button = ({
  data,
  width,
  color,
  columnPos,
  rowPos,
  getBtnData,
}) => {
  const btnStyle = {
    gridColumn: `${columnPos} / span ${width}`,
    gridRow: rowPos,
  };
  return (
    <button
      style={btnStyle}
      className={`btn ${color}`}
      onClick={() => getBtnData(data)}
    >
      {data}
    </button>
  );
};

Button.defaultProps = {
  width: 1,
};

Button.propTypes = {
  data: PropTypes.string.isRequired,
  width: PropTypes.number,
  color: PropTypes.string,
  columnPos: PropTypes.number,
  rowPos: PropTypes.number,
  getBtnData: PropTypes.func.isRequired,
};
