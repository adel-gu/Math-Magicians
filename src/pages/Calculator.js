import { Calculator } from '../components/Calculator';

// Calculator Component
const CalculatorPage = () => (
  <main>
    <div className="container flex justify-between">
      <h1>Lets do some math!</h1>
      <Calculator />
    </div>
  </main>
);

export default CalculatorPage;
