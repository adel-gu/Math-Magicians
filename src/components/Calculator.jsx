import React, { useState } from 'react';
import { Button } from './Button';
import { Screen } from './Screen';
import buttonsData from './buttonData';
import calculate from '../logic/calculate';

// Stylesheets
import '../styles/Calculator.css';

export const Calculator = () => {
  let obj = {
    total: null,
    next: null,
    operation: null,
  };

  const [calObj, setCalcObj] = useState(obj);
  const [result, setResult] = useState('0');

  const getBtnData = (data) => {
    obj = calculate(calObj, data);
    setCalcObj(obj);
    let screenResult =
      `${obj.total ?? ''}${obj.operation ?? ''}${obj.next ?? ''}` || '0';
    setResult(screenResult);
  };

  return (
    <div className="calculator">
      <Screen result={result} />
      {buttonsData.map((btnData) => (
        <Button
          key={`${btnData.data}`}
          data={btnData.data}
          width={btnData.width}
          color={btnData.color}
          columnPos={btnData.columnPos}
          rowPos={btnData.rowPos}
          isOperation={btnData.isOperation}
          getBtnData={getBtnData}
        />
      ))}
    </div>
  );
};
