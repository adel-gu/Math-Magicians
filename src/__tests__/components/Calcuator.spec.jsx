import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { Calculator } from '../../components/Calculator';

beforeEach(() => {
  render(<Calculator />);
});

const calculatorButtonsClicked = (arr) => {
  arr.map((data) => {
    const buttonElement = screen.getByText(data);
    fireEvent.click(buttonElement);
  });
};

describe('Calculator', () => {
  it('should render all calculator components', () => {
    const screenElement = screen.getByTestId('screenTestId');
    const buttonElements = screen.getAllByTestId('buttonTestId');

    expect(screenElement).toBeInTheDocument();
    expect(buttonElements.length).toBe(19);
  });

  it('should set screen initial innerTextContent to "0"', () => {
    const screenElement = screen.getByTestId('screenTestId');

    expect(screenElement.textContent).toBe('0');
  });

  it('should set screen innerTextContent to button data value when buttons are clicked', () => {
    const screenElement = screen.getByTestId('screenTestId');
    calculatorButtonsClicked(['1']);

    expect(screenElement.textContent).toBe('1');
  });

  it('should set screen innerTextContent to button data value when buttons are clicked continuously', () => {
    const screenElement = screen.getByTestId('screenTestId');

    calculatorButtonsClicked(['1', '2', '+']);
    expect(screenElement.textContent).toBe('12+');
  });

  it('should set screen innerTextContent to calculation result when "=" button clicked', () => {
    const screenElement = screen.getByTestId('screenTestId');

    calculatorButtonsClicked(['1', '+', '2', '=']);
    expect(screenElement.textContent).toBe('3');
  });

  it('should set screen innerTextContent to calculation result when "+" button clicked second time after inserting second number', () => {
    const screenElement = screen.getByTestId('screenTestId');

    calculatorButtonsClicked(['1', '+', '2', '+']);
    expect(screenElement.textContent).toBe('3+');
  });

  it('should set screen innerTextContent to "0" when "AC" is clicked', () => {
    const screenElement = screen.getByTestId('screenTestId');

    calculatorButtonsClicked(['1', '+', '2', '=']);
    calculatorButtonsClicked(['AC']);
    expect(screenElement.textContent).toBe('0');
  });

  it('should match snapshot', () => {
    const tree = renderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
