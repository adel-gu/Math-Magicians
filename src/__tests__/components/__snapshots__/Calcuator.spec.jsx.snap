// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator should match snapshot 1`] = `
<div
  className="calculator"
>
  <span
    className="screen"
    data-testid="screenTestId"
  >
    0
  </span>
  <button
    className="btn undefined"
    data-testid="buttonTestId"
    onClick={[Function]}
    style={
      Object {
        "gridColumn": "1 / span 1",
        "gridRow": 2,
      }
    }
  >
    AC
  </button>
  <button
    className="btn undefined"
    data-testid="buttonTestId"
    onClick={[Function]}
    style={
      Object {
        "gridColumn": "2 / span 1",
        "gridRow": 2,
      }
    }
  >
    +/-
  </button>
  <button
    className="btn undefined"
    data-testid="buttonTestId"
    onClick={[Function]}
    style={
      Object {
        "gridColumn": "3 / span 1",
        "gridRow": 2,
      }
    }
  >
    %
  </button>
  <button
    className="btn orang"
    data-testid="buttonTestId"
    onClick={[Function]}
    style={
      Object {
        "gridColumn": "4 / span 1",
        "gridRow": 2,
      }
    }
  >
    +
  </button>
  <button
    className="btn undefined"
    data-testid="buttonTestId"
    onClick={[Function]}
    style={
      Object {
        "gridColumn": "1 / span 1",
        "gridRow": 3,
      }
    }
  >
    7
  </button>
  <button
    className="btn undefined"
    data-testid="buttonTestId"
    onClick={[Function]}
    style={
      Object {
        "gridColumn": "2 / span 1",
        "gridRow": 3,
      }
    }
  >
    8
  </button>
  <button
    className="btn undefined"
    data-testid="buttonTestId"
    onClick={[Function]}
    style={
      Object {
        "gridColumn": "3 / span 1",
        "gridRow": 3,
      }
    }
  >
    9
  </button>
  <button
    className="btn orang"
    data-testid="buttonTestId"
    onClick={[Function]}
    style={
      Object {
        "gridColumn": "4 / span 1",
        "gridRow": 3,
      }
    }
  >
    x
  </button>
  <button
    className="btn undefined"
    data-testid="buttonTestId"
    onClick={[Function]}
    style={
      Object {
        "gridColumn": "1 / span 1",
        "gridRow": 4,
      }
    }
  >
    4
  </button>
  <button
    className="btn undefined"
    data-testid="buttonTestId"
    onClick={[Function]}
    style={
      Object {
        "gridColumn": "2 / span 1",
        "gridRow": 4,
      }
    }
  >
    5
  </button>
  <button
    className="btn undefined"
    data-testid="buttonTestId"
    onClick={[Function]}
    style={
      Object {
        "gridColumn": "3 / span 1",
        "gridRow": 4,
      }
    }
  >
    6
  </button>
  <button
    className="btn orang"
    data-testid="buttonTestId"
    onClick={[Function]}
    style={
      Object {
        "gridColumn": "4 / span 1",
        "gridRow": 4,
      }
    }
  >
    -
  </button>
  <button
    className="btn undefined"
    data-testid="buttonTestId"
    onClick={[Function]}
    style={
      Object {
        "gridColumn": "1 / span 1",
        "gridRow": 5,
      }
    }
  >
    1
  </button>
  <button
    className="btn undefined"
    data-testid="buttonTestId"
    onClick={[Function]}
    style={
      Object {
        "gridColumn": "2 / span 1",
        "gridRow": 5,
      }
    }
  >
    2
  </button>
  <button
    className="btn undefined"
    data-testid="buttonTestId"
    onClick={[Function]}
    style={
      Object {
        "gridColumn": "3 / span 1",
        "gridRow": 5,
      }
    }
  >
    3
  </button>
  <button
    className="btn orang"
    data-testid="buttonTestId"
    onClick={[Function]}
    style={
      Object {
        "gridColumn": "4 / span 1",
        "gridRow": 5,
      }
    }
  >
    /
  </button>
  <button
    className="btn undefined"
    data-testid="buttonTestId"
    onClick={[Function]}
    style={
      Object {
        "gridColumn": "1 / span 2",
        "gridRow": 6,
      }
    }
  >
    0
  </button>
  <button
    className="btn undefined"
    data-testid="buttonTestId"
    onClick={[Function]}
    style={
      Object {
        "gridColumn": "3 / span 1",
        "gridRow": 6,
      }
    }
  >
    .
  </button>
  <button
    className="btn orang"
    data-testid="buttonTestId"
    onClick={[Function]}
    style={
      Object {
        "gridColumn": "4 / span 1",
        "gridRow": 6,
      }
    }
  >
    =
  </button>
</div>
`;
